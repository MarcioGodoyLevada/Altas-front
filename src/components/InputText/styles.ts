import styled, { css, DefaultTheme } from 'styled-components';
import { InputTextProps, IconProps } from '.';

type ContainerProps = {
  error: boolean;
  hasValue: boolean;
};

const modifiers = {
  error: (theme: DefaultTheme) => css`
    ${Label},
    ${IconWrapper}, 
    ${Input} {
      color: ${theme.colors.error};
    }

    ${InputWrapper} {
      border-color: ${theme.colors.error};

      &:focus,
      &:focus-within,
      &:active {
        border-color: ${theme.colors.error};
      }
    }

    ${IconWrapper} {
      cursor: default;
    }
  `,

  floatingLabel: (theme: DefaultTheme, error: boolean) => css`
    font-size: ${theme.font.sizes.xmedium};
    background-image: linear-gradient(
      ${theme.colors.background500},
      ${theme.colors.white}
    );
    transform: scale(0.75) translate(1rem, -170%);

    ${error && modifiers.error(theme)};
  `,
};

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.6rem;
  width: 100%;
`;

export const Container = styled.div<ContainerProps>`
  ${({ theme, error, hasValue }) => css`
    position: relative;
    color: ${theme.colors.primary};

    ${error && modifiers.error(theme)};

    ${Label} {
      ${hasValue && modifiers.floatingLabel(theme, !!error)};
    }

    &:focus-within ${Label} {
      ${modifiers.floatingLabel(theme, !!error)}
    }

    &:focus,
    &:focus-within,
    &:active {
      ${Label} {
        color: ${theme.colors.primary};
      }
    }
  `}
`;

export const InputWrapper = styled.div<InputTextProps>`
  ${({ theme }) => css`
    height: ${theme.input.sizes};
    border: ${theme.border.thickness} solid ${theme.colors.gray300};
    border-radius: ${theme.border.radius};
    display: flex;
    align-items: center;

    &:focus,
    &:focus-within,
    &:active {
      border-color: ${theme.colors.tertiary};
      box-shadow: '0px 0px 0px 4px rgba(99, 131, 123, 0.2)';
    }
  `}
`;

export const Input = styled.input<InputTextProps>`
  ${({ theme, icon }) => css`
    background-color: ${theme.colors.background500};
    border: 0;
    border-radius: ${theme.border.radius};
    color: ${theme.colors.gray500};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding-right: ${icon ? theme.spacings.xxlarge : theme.spacings.xsmall};
    padding-left: ${theme.spacings.xsmall};
    outline: none;
    width: 100%;
    height: 100%;

    &::placeholder {
      color: ${theme.colors.gray500};
    }
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    background-color: transparent;
    color: ${theme.colors.gray500};
    display: inline-block;
    font-size: ${theme.font.sizes.medium};
    font-family: ${theme.font.family};
    line-height: 1.3;
    padding: 0 ${theme.spacings.xxsmall};
    position: absolute;
    top: 50%;
    transform: scale(1) translate(1rem, -50%);
    transition: ${theme.transition.fast};
  `}
`;

export const IconWrapper = styled.div<IconProps>`
  ${({ theme }) => css`
    position: absolute;
    color: ${theme.colors.gray500};
    height: min-content;
    width: 2.2rem;
    right: ${theme.spacings.xsmall};
    cursor: pointer;
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    font-size: ${theme.font.sizes.small};
    overflow: hidden;
    text-overflow: ellipsis;
    width: 100%;
    position: absolute;
    margin-top: 0.4rem;
    text-align: justify;

    #link {
      padding-left: 0.5rem;
    }
  `}
`;

export const Text = styled.span`
  ${({ theme }) => css`
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.secondary};
  `}
`;
