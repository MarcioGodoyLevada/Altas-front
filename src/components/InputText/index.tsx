import React from 'react';

import * as S from './styles';

export type IconProps = {
  icon?: React.ReactElement;
  iconClick?: () => void;
};

export type InputTextProps = {
  className?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  type?: 'text' | 'password';
  value?: string;
  name?: string;
  label?: string;
  placeholder?: string;
  autoFocus?: boolean;
  required?: boolean;
  errorMessage?: string;
  href?: string;
  text?: string;
  maxLength?: number;
} & IconProps;

const InputText = ({
  className,
  onChange,
  name,
  label,
  placeholder,
  value,
  icon,
  autoFocus,
  errorMessage,
  type = 'text',
  href,
  text,
  maxLength,
  iconClick,
  ...props
}: InputTextProps) => {
  return (
    <S.Wrapper className={className}>
      <S.Container error={!!errorMessage} hasValue={!!value || !!placeholder}>
        {!!label && <S.Label htmlFor={name}>{label}</S.Label>}

        <S.InputWrapper>
          <S.Input
            type={type}
            onChange={onChange}
            value={value}
            placeholder={placeholder}
            icon={icon}
            name={name}
            autoFocus={autoFocus}
            maxLength={maxLength}
            {...(label ? { id: name } : {})}
            {...props}
          />

          {!!icon && !errorMessage && (
            <S.IconWrapper onClick={iconClick}>{icon}</S.IconWrapper>
          )}
        </S.InputWrapper>

        {!!errorMessage && <S.Error>{errorMessage}</S.Error>}
      </S.Container>
    </S.Wrapper>
  );
};

export default InputText;
