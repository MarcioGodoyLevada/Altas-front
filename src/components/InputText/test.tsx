import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from '../../utils/tests/helpers';

import InputText from '.';

const LABEL_TRANSITION_AWAITING_TIME = 1000;

describe('<InputText />', () => {
  it('should render with label', () => {
    const { container } = renderWithTheme(
      <InputText label="Input Text" name="field" />
    );

    expect(screen.getByLabelText('Input Text')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without label', () => {
    renderWithTheme(<InputText />);

    expect(screen.queryByLabelText('Input Text')).not.toBeInTheDocument();
  });

  it('should render label correctly when input is focused', async () => {
    renderWithTheme(<InputText label="Input Text" name="field" />);

    userEvent.tab();

    setTimeout(() => {
      expect(screen.getByText('Input Text')).toHaveStyle({
        transform: 'scale(1) translate(1rem,-50%)',
      });
    }, LABEL_TRANSITION_AWAITING_TIME);
  });

  it('should render label correctly when input is filled', async () => {
    renderWithTheme(
      <InputText label="Input Text" name="field" value="value" />
    );

    setTimeout(() => {
      expect(screen.getByText('Input Text')).toHaveStyle({
        transform: 'scale(1) translate(1rem,-50%)',
      });
    }, LABEL_TRANSITION_AWAITING_TIME);
  });

  it('should render with placeholder', () => {
    renderWithTheme(<InputText placeholder="Placeholder Text" />);

    expect(screen.getByPlaceholderText('Placeholder Text')).toBeInTheDocument();
  });

  it('it should change its value when typing', async () => {
    const onChange = jest.fn();

    renderWithTheme(
      <InputText onChange={onChange} label="Input Text" name="field" />
    );

    const input = screen.getByRole('textbox');
    const text = 'My new text';

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onChange).toHaveBeenCalledTimes(text.length);
    });
  });

  it('should be accessible by tab', () => {
    renderWithTheme(<InputText label="Input Text" name="field" />);

    const input = screen.getByLabelText('Input Text');

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).toHaveFocus();
  });

  it('should render with error', () => {
    const { container } = renderWithTheme(<InputText errorMessage="Error" />);

    expect(screen.getByText(/error/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
