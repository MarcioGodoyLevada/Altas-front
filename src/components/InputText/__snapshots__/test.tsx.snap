// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputText /> should render with error 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.6rem;
  width: 100%;
}

.c1 {
  position: relative;
}

.c1 .c2 {
  border-color: #FDA29B;
}

.c1 .c2:focus,
.c1 .c2:focus-within,
.c1 .c2:active {
  border-color: #FDA29B;
}

.c1 .c6 {
  cursor: default;
}

.c3 {
  border-radius: 0.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 4.2rem;
  border: 0.2rem solid #D0D5DD;
  border-color: #9CB1BC;
}

.c3:focus,
.c3:focus-within,
.c3:active {
  box-shadow: 0px 0px 0px 4px rgba(99,131,123,0.2);
}

.c3 .c4 {
  background-color: #F2F3F4;
}

.c5 {
  border: 0;
  border-radius: 0.8rem;
  color: #3C3E3E;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Lato,Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-size: 1.6rem;
  padding-right: 0.8rem;
  padding-left: 0.8rem;
  outline: none;
  width: 100%;
  height: 100%;
}

.c5::-webkit-input-placeholder {
  color: #3C3E3E;
}

.c5::-moz-placeholder {
  color: #3C3E3E;
}

.c5:-ms-input-placeholder {
  color: #3C3E3E;
}

.c5::placeholder {
  color: #3C3E3E;
}

.c7 {
  position: absolute;
  color: #667085;
  height: -webkit-min-content;
  height: -moz-min-content;
  height: min-content;
  width: 2.2rem;
  right: 0.8rem;
  cursor: pointer;
}

.c8 {
  color: #F04438;
  font-size: 1.4rem;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 100%;
  position: absolute;
  margin-top: 0.4rem;
  text-align: justify;
}

.c8 #link {
  padding-left: 0.5rem;
}

.c9 {
  color: #3F534E;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Lato,Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-size: 1.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c9:active {
  font-weight: 300;
}

.c9:hover,
.c9:focus {
  opacity: 0.8;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2 c3"
      sizes="medium"
    >
      <input
        class="c4 c5"
        type="text"
        value=""
      />
      <div
        class="c6 c7"
      >
        <svg
          fill="none"
          height="22"
          viewBox="0 0 22 22"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7V11M11 15H11.01M21 11C21 16.5228 16.5228 21 11 21C5.47715 21 1 16.5228 1 11C1 5.47715 5.47715 1 11 1C16.5228 1 21 5.47715 21 11Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <p
      class="c8"
    >
      Error
      <a
        class="c9"
        href=""
        id="link"
        rel="noreferrer noopener"
      />
    </p>
  </div>
</div>
`;

exports[`<InputText /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.6rem;
  width: 100%;
}

.c1 {
  position: relative;
}

.c1:focus-within .c2 {
  display: block;
  color: #3B5249;
}

.c1:focus .c2,
.c1:focus-within .c2,
.c1:active .c2 {
  color: #3B5249;
}

.c4 {
  border-radius: 0.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 4.2rem;
  border: 0.2rem solid #D0D5DD;
  border-color: #9CB1BC;
}

.c4:focus,
.c4:focus-within,
.c4:active {
  box-shadow: 0px 0px 0px 4px rgba(99,131,123,0.2);
}

.c4 .c5 {
  background-color: #F2F3F4;
}

.c6 {
  border: 0;
  border-radius: 0.8rem;
  color: #3C3E3E;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Lato,Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-size: 1.6rem;
  padding-right: 0.8rem;
  padding-left: 0.8rem;
  outline: none;
  width: 100%;
  height: 100%;
}

.c6::-webkit-input-placeholder {
  color: #3C3E3E;
}

.c6::-moz-placeholder {
  color: #3C3E3E;
}

.c6:-ms-input-placeholder {
  color: #3C3E3E;
}

.c6::placeholder {
  color: #3C3E3E;
}

.c3 {
  background-color: transparent;
  color: #667085;
  display: inline-block;
  font-size: 1.6rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Lato,Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.3;
  padding: 0 0.4rem;
  position: absolute;
  top: 50%;
  -webkit-transform: scale(1) translate(1rem,-50%);
  -ms-transform: scale(1) translate(1rem,-50%);
  transform: scale(1) translate(1rem,-50%);
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2 c3"
      for="field"
    >
      Input Text
    </label>
    <div
      class="c4"
      sizes="medium"
    >
      <input
        class="c5 c6"
        id="field"
        name="field"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
